src/main/java/io/hhplus/tdd/point/facade/PointFacade.java

```diff
@@ -0,0 +1,100 @@
+package io.hhplus.tdd.point.facade;
+
+import io.hhplus.tdd.point.dto.PointHistory;
+import io.hhplus.tdd.point.dto.UserPoint;
+import io.hhplus.tdd.point.service.PointHistoryService;
+import io.hhplus.tdd.point.service.PointService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+
+@Service
+@RequiredArgsConstructor
+public class PointFacade {
+    private final PointService pointService;
+    private final PointHistoryService pointHistoryService;
+
+    // 사용자별 작업 큐 및 Lock 관리
+    private final ConcurrentHashMap<Long, Queue<Runnable>> userQueues = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<Long, ReentrantLock> userLocks = new ConcurrentHashMap<>();
+
+    // 사용자별 Lock 가져오기
+    private ReentrantLock getUserLock(long userId) {
+        return userLocks.computeIfAbsent(userId, id -> new ReentrantLock());
+    }
+
+    // 사용자별 Queue 가져오기
+    private Queue<Runnable> getUserQueue(long userId) {
+        return userQueues.computeIfAbsent(userId, id -> new LinkedList<>());
+    }
+
+    // 작업 큐에 작업 추가 및 처리 시작
+    private void submitTask(long userId, Runnable task) {
+        Queue<Runnable> queue = getUserQueue(userId);
+
+        synchronized (queue) {
+            queue.add(task); // 작업을 큐에 추가
+            if (queue.size() == 1) {
```

Comment:
```suggestion
            if (queue.size() == 1) {
```

이 부분의 의미가 명확하지 않아 1이 갖는 의미와 상황을 잘 모르겠네요.
+ 해당 코드 라인 바로 위에는 add() 를 호출하니 항상 1이 아닌가요? (어떤 케이스가 염두된건지 궁금합니다.)

Reply:
일단 queue의 size로 조건을 건 부분은 
여러 스레드가 동일한 `userId`로 호출될 때, 작업이 동시에 큐에 추가되는 경우를 방지하기 위한 로직입니다. 

예를 들어서, 두개의 스레드가 동시에 `queue.add(task)`를 실행하면, 첫번째 스레드가 `processQueue`를 호출하기 전에 두번째 작업이 큐에 추가 될 경우에 queue.size()가 2 이상이 될 수 있기 때문에 
새로운 작업이 추가될 때마다 `processQueue`가 호출되어서 중복 실행 문제가 발생되더라구요.

`processQueue`가 이미 실행중라면 큐에서 이미 순차적으로 처리하기 때문에 
`queue.add(task)`로 추가된 이후에도 큐 크기는 1 이상이 되고, 새로운 스레드를 생성할 필요가 없다고 생각했습니다.

그래서 `queue.size()`가 항상 1일 때만 `processQueue`를 호출하도록 제어하면 큐에 있는 기존 작업과 추가 작업이 하나의 스레드에서 처리하도록 보장할 수 있다고 생각하여 추가했습니다!

---

요약하자면, 아래와 같이 생각해주시면 될 것 같습니다.
1. **중복 실행 방지** : 동일한 사용자 큐에서 작업이 실행중일 때 추가 스레드가 불필요하게 생성되지 않음
2. **순서보장(FIFO)** : 큐에 추가된 작업을 하나의 스레드에서 순차적으로 처리해서 순서 보장

---

아마 제가 해당 부분을 테스트코드로 작성하지 못했고, 별도의 주석도 없어서 더 이해 안가셨을 것 같습니다..ㅠㅠ

질문 감사합니다~!